### ISSUE TITLE
Mypy allows to compare NewTypes created from Decimal 

### ISSUE THREAD

**Issue by kulych** [python/mypy#15095]
**Bug Report**

Mypy does not produce an error when comparing `NewType`s based on `Decimal`. 

**To Reproduce**

```python
from decimal import Decimal
from typing import NewType

# Comparing Decimal NewTypes is OK 
FirstDecimal = NewType("FirstDecimal", Decimal)
SecondDecimal = NewType("SecondDecimal", Decimal)

FirstDecimal(Decimal("0.1")) == SecondDecimal(Decimal("0.1"))  # ! mypy ok

# But comparing other NewTypes is not
FirstInt = NewType("FirstInt", int)
SecondInt = NewType("SecondInt", int)

FirstInt(1) == SecondInt(3)  # mypy error

FirstString = NewType("FirstString", str)
SecondString = NewType("SecondString", str)

FirstString("bam") == SecondString("boo")  # mypy error
```

**Expected Behavior**

Mypy produces `Non-overlapping equality check` for all three comparisons.

**Actual Behavior**

Mypy produces `Non-overlapping equality check` only for `int` and `str` NewTypes.

**Your Environment**

- Mypy version used: 1.1.1
- Mypy command-line flags: `--strict`
- Mypy configuration options from `mypy.ini` (and other config files):
```
[tool.mypy]
strict = true
show_error_context = true
show_error_codes = true
```
- Python version used: `3.10.8`

**Comment by erictraut** [comment]
I'm surprised that mypy would generate `comparison-overlap` errors for any of these cases. As I understand it, `comparison-overlap` is intended to detect cases where a comparison will always produce `True` or `False` because the types of the operands are not comparable. This is how the feature is [documented](https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-comparisons-are-overlapping-comparison-overlap).

In the three cases above, the operand types _are_ comparable, and they can produce either `True` or `False` depending on the values.

```python
print(FirstDecimal(Decimal("0.1")) == SecondDecimal(Decimal("0.2"))) # False
print(FirstDecimal(Decimal("0.1")) == SecondDecimal(Decimal("0.1"))) # True

print(FirstString("bam") == SecondString("boo")) # False
print(FirstString("boo") == SecondString("boo")) # True
```

So, I think mypy shouldn't produce `comparison-overlap` errors for any of these cases.

**Comment by tk-ematiq** [comment]
Just to improve my provided example:
```python3
from decimal import Decimal
from typing import NewType

# Comparing Decimal NewTypes is OK 
FirstDecimal = NewType("FirstDecimal", Decimal)
SecondDecimal = NewType("SecondDecimal", Decimal)

result_decimal_false = FirstDecimal(Decimal("0.6")) == SecondDecimal(Decimal("0.1"))  # ! mypy ok
result_decimal_true = FirstDecimal(Decimal("0.1")) == SecondDecimal(Decimal("0.1"))  # ! mypy ok
print("Decimal false:", result_decimal_false)
print("Decimal true:", result_decimal_true)
print()

# But comparing other NewTypes is not
FirstInt = NewType("FirstInt", int)
SecondInt = NewType("SecondInt", int)

result_int_false = FirstInt(1) == SecondInt(3)  # mypy error
result_int_true = FirstInt(1) == SecondInt(1)  # mypy error
print("Float false:", result_int_false)
print("Float true:", result_int_true)
print()

FirstString = NewType("FirstString", str)
SecondString = NewType("SecondString", str)

result_string_false = FirstString("bam") == SecondString("boo")  # mypy error
result_string_true = FirstString("bam") == SecondString("bam")  # mypy error

print("String false:", result_string_false)
print("String true:", result_string_true)
```

Produces (as @erictraut correctly noted)
```
Decimal false: False
Decimal true: True

Float false: False
Float true: True

String false: False
String true: True
```

And mypy output is:
```
(venv) [~/tmp]$ mypy --strict test_mypy.py
test_mypy.py:18: error: Non-overlapping equality check (left operand type: "FirstInt", right operand type: "SecondInt")  [comparison-overlap]
test_mypy.py:19: error: Non-overlapping equality check (left operand type: "FirstInt", right operand type: "SecondInt")  [comparison-overlap]
test_mypy.py:27: error: Non-overlapping equality check (left operand type: "FirstString", right operand type: "SecondString")  [comparison-overlap]
test_mypy.py:28: error: Non-overlapping equality check (left operand type: "FirstString", right operand type: "SecondString")  [comparison-overlap]
Found 4 errors in 1 file (checked 1 source file)
```

I am in no means a `mypy` expert so please take this issues only as an inconsistency report.

**Comment by kulych** [comment]
Just to improve my provided example:
```python3
from decimal import Decimal
from typing import NewType

# Comparing Decimal NewTypes is OK 
FirstDecimal = NewType("FirstDecimal", Decimal)
SecondDecimal = NewType("SecondDecimal", Decimal)

result_decimal_false = FirstDecimal(Decimal("0.6")) == SecondDecimal(Decimal("0.1"))  # ! mypy ok
result_decimal_true = FirstDecimal(Decimal("0.1")) == SecondDecimal(Decimal("0.1"))  # ! mypy ok
print("Decimal false:", result_decimal_false)
print("Decimal true:", result_decimal_true)
print()

# But comparing other NewTypes is not
FirstInt = NewType("FirstInt", int)
SecondInt = NewType("SecondInt", int)

result_int_false = FirstInt(1) == SecondInt(3)  # mypy error
result_int_true = FirstInt(1) == SecondInt(1)  # mypy error
print("Float false:", result_int_false)
print("Float true:", result_int_true)
print()

FirstString = NewType("FirstString", str)
SecondString = NewType("SecondString", str)

result_string_false = FirstString("bam") == SecondString("boo")  # mypy error
result_string_true = FirstString("bam") == SecondString("bam")  # mypy error

print("String false:", result_string_false)
print("String true:", result_string_true)
```

Produces (as @erictraut correctly noted)
```
Decimal false: False
Decimal true: True

Float false: False
Float true: True

String false: False
String true: True
```

And mypy output is:
```
(venv) [~/tmp]$ mypy --strict test_mypy.py
test_mypy.py:18: error: Non-overlapping equality check (left operand type: "FirstInt", right operand type: "SecondInt")  [comparison-overlap]
test_mypy.py:19: error: Non-overlapping equality check (left operand type: "FirstInt", right operand type: "SecondInt")  [comparison-overlap]
test_mypy.py:27: error: Non-overlapping equality check (left operand type: "FirstString", right operand type: "SecondString")  [comparison-overlap]
test_mypy.py:28: error: Non-overlapping equality check (left operand type: "FirstString", right operand type: "SecondString")  [comparison-overlap]
Found 4 errors in 1 file (checked 1 source file)
```

I am in no means a `mypy` expert so please take this issues only as an inconsistency report.

**Comment by ikonst** [comment]
`comparison-overlap`, as defined, indeed seems incorrect as the thing that flags NewType violations. We should either rewrite the document or make a new error code. The validation, however, seems valid, as the point of having a NewType is to make the checker yell at you for, say, interchanging miles and meters.
